src/mesh/rbf.f90

```diff
@@ -0,0 +1,597 @@
+! Upper left part: M_i,j = phi(abs(X(i)-X(j)))
+DO j=1,nBP; DO i=1,nBP
+  dist = Distance(RefCoordinates(:,i),RefCoordinates(:,j))
+  RBFMatrix(i,j) = EvaluateRBF(dist,SupportRadius(iRBFBox),RBFType(iRBFBox))
+END DO; END DO ! j,i=1,nBP
+! Upper right part: P_b - row i equal to [1 x_i y_i z_i]
+! Lower left part: transpose of P_b
+DO i=1,nBP
+   RBFMatrix(i,nBP+1) = 1.
+   RBFMatrix(nBP+1,i) = 1.
+   RBFMatrix(i,nBP+2) = RefCoordinates(1,i)
+   RBFMatrix(nBP+2,i) = RefCoordinates(1,i)
+   RBFMatrix(i,nBP+3) = RefCoordinates(2,i)
+   RBFMatrix(nBP+3,i) = RefCoordinates(2,i)
+   RBFMatrix(i,nBP+4) = RefCoordinates(3,i)
+   RBFMatrix(nBP+4,i) = RefCoordinates(3,i)
+END DO ! i=1,nBP
+! Lower right part: zeros
+DO j=nBP+1,nBP+4; DO i=nBP+1,nBP+4
+  RBFMatrix(i,j) = 0.
+END DO; END DO ! j,i=nBP+1,nBP+4
+
+! Solve the system for all RHSs - RBFRHS will then contain the solution to the system!
+CALL SolveLinSys(nBP+4,3,RBFMatrix,RBFRHS)
+DEALLOCATE(RBFMatrix)
+
+!--------------------------------------------------------------------------------------------------------!
+! Evaluate the interpolation for all nodes in the volume
+!--------------------------------------------------------------------------------------------------------!
+WRITE(*,*) 'Evaluating RBF interpolation... '
+
+Elem=>FirstElem
+DO WHILE(ASSOCIATED(Elem))
+  ! Calculate the tri-linear coordinates of the curved nodes (those are all control points of the RBF interpolation, so they don't
+  ! change due to the displacement procedure)
+  xCornerVol(:,HexaMapInvLinear(0,0,0)) = Elem%CurvedNode(HexaMapInv(0,0,0))%np%x
+  xCornerVol(:,HexaMapInvLinear(0,1,0)) = Elem%CurvedNode(HexaMapInv(0,N,0))%np%x
+  xCornerVol(:,HexaMapInvLinear(1,0,0)) = Elem%CurvedNode(HexaMapInv(N,0,0))%np%x
+  xCornerVol(:,HexaMapInvLinear(1,1,0)) = Elem%CurvedNode(HexaMapInv(N,N,0))%np%x
+  xCornerVol(:,HexaMapInvLinear(0,0,1)) = Elem%CurvedNode(HexaMapInv(0,0,N))%np%x
+  xCornerVol(:,HexaMapInvLinear(0,1,1)) = Elem%CurvedNode(HexaMapInv(0,N,N))%np%x
+  xCornerVol(:,HexaMapInvLinear(1,0,1)) = Elem%CurvedNode(HexaMapInv(N,0,N))%np%x
+  xCornerVol(:,HexaMapInvLinear(1,1,1)) = Elem%CurvedNode(HexaMapInv(N,N,N))%np%x
+  ! Get a guess for the center of the current element by averaging the corner nodes
+  xBary(:) = Elem%CurvedNode(HexaMapInv(0,0,0))%np%x &
+           + Elem%CurvedNode(HexaMapInv(0,N,0))%np%x &
+           + Elem%CurvedNode(HexaMapInv(N,0,0))%np%x &
+           + Elem%CurvedNode(HexaMapInv(N,N,0))%np%x &
+           + Elem%CurvedNode(HexaMapInv(0,0,N))%np%x &
+           + Elem%CurvedNode(HexaMapInv(0,N,N))%np%x &
+           + Elem%CurvedNode(HexaMapInv(N,0,N))%np%x &
+           + Elem%CurvedNode(HexaMapInv(N,N,N))%np%x
+  xBary(:) = xBary(:) / 8.
+  ! Cycle if the bary center is not inside of the bounding box (thus only all or no points of an element are curved)
+  IF ((xBary(1).LT.xMin).OR.(xBary(1).GT.xMax).OR.(xBary(2).LT.yMin).OR.(xBary(2).GT.yMax)) THEN
+    Elem=>Elem%nextElem
+    CYCLE
+  END IF
```

Comment:
So, we compute the RBF on all nodes within the bounding box but only apply it to whole elements. Is this more consistent than excluding the elements beforehand?

Reply:
Yes, should be selected by checking whether the bounding box of the linear element and the RBF domain are overlapping. Then we evaluate the RBF basis for this element.